### ISSUE TITLE
Bio.codonalign status

### ISSUE THREAD

**Issue by mdehoon** [biopython/biopython#3478]
`Bio.codonalign` was added to Biopython about 8 years ago. Currently this module is still considered to be experimental, and prints a `BiopythonExperimentalWarning` when imported:

```python
>>> from Bio import codonalign
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/biopython-1.79.dev0-py3.6-macosx-10.9-x86_64.egg/Bio/codonalign/__init__.py:26: BiopythonExperimentalWarning: Bio.codonalign is an experimental module which may undergo significant changes prior to its future official release.
  BiopythonExperimentalWarning,
>>> 
```

We should reach a decision whether this module can be officially released. This would require integration of the `codonalign` documentation (which is currently in a separate LaTeX file that is not compiled or distributed) into the Biopython tutorial, and to revisit `Bio.codonalign` to see how much of its functionality is already provided (or can better be provided) by existing Biopython modules.

**Comment by mdehoon** [comment]
@zruan

**Comment by zruan** [comment]
I agree that it is a good idea to officially release the module. The basic function of `Bio.codonalign` is sufficiently stable and many bugs have been fixed over the past few years. The tutorial needs to be revisited to ensure it reflects the latest changes of the code.

For the functionality check, I have to admit that I've not been working on analyzing codon alignment data for quite some time and not aware of the best practice in the field. The current implementation of dN/dS calculation methods and MK test is preliminary and worth looking by an extra eye.

**Comment by mdehoon** [comment]
@zruan 
Thank you.
I am still not sure though what functionality `Bio.codonalign` provides, and how much of that functionality is already available in other Biopython modules.

**Comment by zruan** [comment]
Hi @mdehoon,

The main function that `Bio.codonalign` provides is to generate alignment of codons based on a protein sequence alignment and the corresponding DNA sequences that encode the protein sequences. I don't think this function is available in other Biopython modules. The module also has some basic auxiliary functions such as counting the number of synonymous and non-synonymous substitutions which is also unique to the module.

**Comment by mdehoon** [comment]
@zruan  I guess I am trying to understand for what purpose a user would want to generate a codon alignment based on a protein sequence alignment. I assume that the user would start from a set of nucleotide sequences? And the user translates them to protein sequences, and aligns those? And then `Bio.codonalign` uses the protein sequence alignment to get the nucleotide sequence alignment on a codon-by-codon basis?

**Comment by zruan** [comment]
@mdehoon Yes. this is usually the case. This type of analysis used to be very popular to study the gene evolution within a population.

**Comment by peterjc** [comment]
Yes, I wrote a Python script to specifically "back-translate a protein alignment to nucleotides" way back in 2012:

https://github.com/peterjc/picobio/commit/f2efdda4641cf60342facc3b774c089c211c9ddb

I later added a Galaxy wrapper:

https://github.com/peterjc/pico_galaxy/tree/master/tools/align_back_trans

It needs to be updated for the no-alphabet changes in Biopython, but the use case remains.

**Comment by santiagosnchez** [comment]
I also have a Python wrapper that works pretty well and I haven't detected any issues so far.

https://github.com/santiagosnchez/CodonAlign

**Comment by peterjc** [comment]
This module is still marked as experimental as of Biopython 1.81:

https://github.com/biopython/biopython/blob/biopython-181/Bio/codonalign/__init__.py#L23

If we do remove that status, we ought to include the CodonAlign chapter in the Tutorial, currently in LaTeX and RST format here:

* https://github.com/biopython/biopython/blob/biopython-181/Doc/doc.tex
* https://github.com/biopython/biopython/blob/biopython-181/Doc/doc.rst

This needs at least the "Chapter X" bits removed (and probably we can discard the RST version although it might be useful for sanity testing any pandoc conversion as part of #907).

**Comment by mdehoon** [comment]
I think the `Bio.codonalign` module is effectively obsolete; this functionaliyt can be handled better by `Bio.Align`. I can write a short section for the Tutorial outlining how to use `Bio.Align` for codon alignments.

**Comment by peterjc** [comment]
That sounds reasonable to me, and hopefully @zruan can review it to confirm it covers his original motivating use cases :)

**Comment by zruan** [comment]
@mdehoon I'm happy to review the tutorial section of using `Bio.Align` module to handle codon alignment. The most important function is `Bio.codonalign.build()`, where an input of a protein alignment and the corresponding nucleotide sequences can be converted to a codon-based alignment. There are other auxiliary functions (cal_dn_ds, mktest etc) that may be ported to Bio.Align module for analyzing codon alignment.

**Comment by mdehoon** [comment]
@zruan  Something like this?

```python
>>> from Bio.Seq import translate
>>> from Bio.Align import PairwiseAligner
>>> dna1 = "ATGCAAAAATGCTGGAGT"
>>> dna2 = "ATGAAATGCTGGAGT"
>>> prot1 = translate(dna1)
>>> prot2 = translate(dna2)
>>> aligner = PairwiseAligner()
>>> alignments = aligner.align(prot1, prot2)
>>> len(alignments)
1
>>> alignment = alignments[0]
>>> print(alignment)  # protein alignment
target            0 MQKCWS 6
                  0 |-|||| 6
query             0 M-KCWS 5

>>> alignment.coordinates *= 3
>>> alignment.sequences[0] = dna1
>>> alignment.sequences[1] = dna2
>>> print(alignment)  # codon alignment
target            0 ATGCAAAAATGCTGGAGT 18
                  0 |||---|||||||||||| 18
query             0 ATG---AAATGCTGGAGT 15

```

**Comment by peterjc** [comment]
That's surprisingly short, and elegant that you can just multiple the coordinates by three. Very nice.

I wonder how easy it would be to convert the codon alignment analysis functions to work with this kind of structure? Probably rather more work...

**Comment by mdehoon** [comment]
> I wonder how easy it would be to convert the codon alignment analysis functions to work with this kind of structure? Probably rather more work...

Some of that functionality is already available from `Bio.SeqUtils`, e.g. the `CodonAdaptationIndex` class.

**Comment by zruan** [comment]
@mdehoon Are you able to handle multiple sequence alignment as opposed to pairwise sequence alignment? The protein sequence alignment is typically handled by external software, rather than within biopython. Test data can be found at `Tests/codonalign/nucl1.fa` and `Tests/codonalign/prot1.aln`.

**Comment by mdehoon** [comment]
> @mdehoon Are you able to handle multiple sequence alignment as opposed to pairwise sequence alignment? The protein sequence alignment is typically generated by external software, rather than within biopython. Test data can be found at `Tests/codonalign/nucl1.fa` and `Tests/codonalign/prot1.aln`.

Yes, it works the same way:
```python
>>> from Bio import Align
>>> from Bio import SeqIO

>>> alignment = Align.read("pro1.aln", "clustal")
>>> print(alignment)
isotig697         0 RGDQRSNFQLSPSTMQISTGLLCLLLVATGFTSQVLAHPGSI--PSTYCFVMTSKKIPKS
ENSG00000         0 --------------MKVSAALLCLLLIAATFIPQGLAQPDAINAPVTCCYNFTNRKISVQ
ENSG00000         0 --------------MKVSAALLCLLLIAATFIPQGLAQPDAINAPVTCCYNFTNRKISVQ

isotig697        58 LLKSYKRISNSRCTLKAILFKTKSGKEICADPKKKWVQDATKHLDQILQTPKPTIPSFET
ENSG00000        46 RLASYRRITSSKCPKEAVM-----------------------------------------
ENSG00000        46 RLASYRRITSSKCPKEAVIFKTIVAKEICADPKQKWVQDSMDHLDKQTQTPKT-------

isotig697       118 HPETKKCFIHSPFLRRAPRSTQHHSPRTWLHLVMDRTESHYVQNKPDLKRLCNFLNMQNL
ENSG00000        65 ------------------------------------------------------------
ENSG00000        99 ------------------------------------------------------------

isotig697       178 KRGAC 183
ENSG00000        65 -----  65
ENSG00000        99 -----  99

>>> sequences = SeqIO.parse("nucl1.fa", "fasta")
>>> alignment.sequences = list(sequences)
>>> alignment.coordinates *= 3
>>> print(alignment)
isotig697         0 AGAGGCGATCAACGCAGCAACTTCCAGCTGTCTCCCTCCACCATGCAGATCTCCACAGGG
ENSG00000         0 ------------------------------------------ATGAAAGTCTCTGCCGCC
ENSG00000         0 ------------------------------------------AAAAAAAAAATGAAAGTC

isotig697        60 CTTCTGTGCcTGCTGCTTGTGGCCACTGGCTTCACTTCCCAGGTGCTGGCTCACCCAGGC
ENSG00000        18 CTTCTGTGCCTGCTGCTCATAGCAGCCACCTTCATTCCCCAAGGGCTCGCTCAGCCAGAT
ENSG00000        18 TCTGCCGCCCTTCTGTGCCTGCTGCTCATAGCAGCCACCTTCATTCCCCAAGGGCTCGCT

isotig697       120 TCTATC------CCATCTACCTaCTGCTTTGTTATGACCAGTAAGAaGATCCCCAAATCA
ENSG00000        78 GCAATCAATGCCCCAGTCACCTGCTGTTATAACTTCACCAATAGGAAGATCTCAGTGCAG
ENSG00000        78 CAGCCAGATGCAATCAATGCCCCAGTCACCTGCTGTTATAACTTCACCAATAGGAAGATC

isotig697       174 CTACTGAaGAGCTACAAAaGAATCTCCAACAGCAGATGCACCcTGAAAGCCATACTCTTC
ENSG00000       138 AGGCTCGCGAGCTATAGAAGAATCACCAGCAGCAAGTGTCCCAAAGAAGCTGTGATG---
ENSG00000       138 TCAGTGCAGAGGCTCGCGAGCTATAGAAGAATCACCAGCAGCAAGTGTCCCAAAGAAGCT

isotig697       234 AAGACCAAGTCGGGCAAAGAGATCTGTGCTGACCCCAAGAAGAAGTGGGTCcAGGATGCC
ENSG00000       195 ------------------------------------------------------------
ENSG00000       198 GTGATCTTCAAGACCATTGTGGCCAAGGAGATCTGTGCTGACCCCAAGCAGAAGTGGGTT

isotig697       294 ACAAAGCACCTGGACCAAATCCTTCAAACTCCAAAACCGACAATCCCCTCTTTTGAGACT
ENSG00000       195 ------------------------------------------------------------
ENSG00000       258 CAGGATTCCATGGACCACCTGGACAAGCAAACCCAAACT---------------------

isotig697       354 CACCCAGAGACTAAGAAATGCTTCATTCATTCTCCATTCCTAAGACGTGCTCCAAGGTCA
ENSG00000       195 ------------------------------------------------------------
ENSG00000       297 ------------------------------------------------------------

isotig697       414 ACTCAGCACCATTCCCCAAGGACTTGGCTTCATTTAGTTATGGATAGAACTGAAAGTCAT
ENSG00000       195 ------------------------------------------------------------
ENSG00000       297 ------------------------------------------------------------

isotig697       474 TATGTTCAGAATAAGCCAGACTTGAAGAGGTTGTGTAATTTCTTGAATATGCAAAATCTT
ENSG00000       195 ------------------------------------------------------------
ENSG00000       297 ------------------------------------------------------------

isotig697       534 AAAAGGGGGGCATGC 549
ENSG00000       195 --------------- 195
ENSG00000       297 --------------- 297

```

**Comment by zruan** [comment]
This is great! The `Bio.codonalign.build()` is more rigorous as it can also handle mismatches and non-coding sequences. For example:

```
from Bio.SeqRecord import SeqRecord
from Bio.Seq import Seq
from Bio.Align import MultipleSeqAlignment
from Bio.codonalign import build

seq3 = SeqRecord(Seq("AAAAAAAAAAAAAAAAAAAAAAAAAAATGAAAAAGGAGTTTTTTTTTTTTTTTTTTTTT"), id="pro1")
seq4 = SeqRecord(Seq("CCCCCCCCCCCCCCCCCCCCCCCCCCATGAACAAGCACGAGTTTTTTTTTTTTTTTTTT"), id="pro2")
seq5 = SeqRecord(Seq("CCCCCCCCCCCCCCCCCCCCCCCCCCATGAAAAAGGCTGAGTTTTTTTTTTTTTTTTTT"), id="pro3")
pro3 = SeqRecord(Seq("MKK-E"), id="pro1")
pro4 = SeqRecord(Seq("MNKHE"), id="pro2")
pro5 = SeqRecord(Seq("MKKAE"), id="pro3")
aln = MultipleSeqAlignment([pro3, pro4, pro5])
seqlist = [seq3, seq4, seq5]
a = build(aln, seqlist)
print(a)

CodonAlignment with 3 rows and 15 columns (5 codons)
ATGAAAAAG---GAG pro1
ATGAACAAGCACGAG pro2
ATGAAAAAGGCTGAG pro3
```

This might be helpful for people working on RNAseq data where non-coding region could be present in mRNA/nucleotide sequences.

If we decide to drop the codonalign module, it's probably a good idea to have a function similar to `codonalign.build()` somewhere in biopython so that it won't break the pipeline that others may already have.

**Comment by mdehoon** [comment]
@zruan  Thank you for the detailed example.

I just have some difficulty in understanding how the `Bio.codonalign.build` function is used in practice.
I guess users would start from the sequences `seq3`, `seq4`, and `seq5`. But then they would have to generate `pro3`, `pro4`, and `pro5` by themselves? That would mean translating `seq3`, `seq4`, and `seq5`, and deciding where the protein alignment starts and ends (and which part of the sequence is non-coding). Somewhere along this process, it may be trivial to create the corresponding codon alignment.

**Comment by zruan** [comment]
What you described is generally true. However, there are situations where users may want to focus on the protein alignment instead of nucleotide sequences. For example, if only a certain domain of the protein is of interest, then the users will just need to edit the protein alignment to cover regions of interest without having to worry about the nucleotide sequence. The condoalign.build function close resembles that if [Pal2nal](http://www.bork.embl.de/pal2nal/) where some corner cases such as mismatches, frameshift, and peusodgenes can be handled.

**Comment by mdehoon** [comment]
@zruan Is `Bio.codonalign` essentially a translation of the core parts of pal2nal to Python? Some of the code in `Bio.codonalign` and `pal2nal.pl` looks quite similar. I am asking because `pal2nal.pl` was released under the GPL license.

**Comment by peterjc** [comment]
@santiagosnchez have you had a chance to see how easy it would be to update https://github.com/santiagosnchez/CodonAlign to use ``Bio.Align`` (from a recent Biopython) in place of ``Bio.codonalign``?

**Comment by zruan** [comment]
> _If_ that is the case (GPL violation), it would be an even stronger reason to drop the experimental `Bio.codonalign` in favour of the more recent work in `Bio.Align`.



> @zruan Is `Bio.codonalign` essentially a translation of the core parts of pal2nal to Python? Some of the code in `Bio.codonalign` and `pal2nal.pl` looks quite similar. I am asking because `pal2nal.pl` was released under the GPL license.

The basic concept between the codonalign.build method and pal2nal is similar. Both programs reply on constructing the regular expression from protein sequence alignment to anchor the nucleotide sequences.

**Comment by mdehoon** [comment]
@zruan See #4479 for an independent implementation of a codon aligner using dynamic programming instead of regular expressions.
